movement/validators/datasets.py

```diff
@@ -43,42 +43,212 @@ def _convert_fps_to_none_if_invalid(fps: float | None) -> float | None:
-):
-    """Raise a ValueError if the list does not have the expected length."""
-    if (value is not None) and (len(value) != expected_length):
-        raise logger.error(
-            ValueError(
-                f"Expected '{attribute.name}' to have "
-                f"length {expected_length}, but got {len(value)}."
+    @confidence_array.validator
+    def _validate_confidence_array(self, attribute, value):
+        """Check confidence_array type and shape."""
+        if value is not None:
+            expected_shape = self._confidence_expected_shape
+            self._validate_array_shape(
+                attribute, value, expected_shape=expected_shape
             )
-        )
+
+    @individual_names.validator
+    def _validate_individual_names(self, attribute, value):
+        """Validate individual_names length and uniqueness."""
+        if value is not None:
+            individuals_dim_index = self.DIM_NAMES.index("individuals")
+            self._validate_list_length(
+                attribute,
+                value,
+                self.position_array.shape[individuals_dim_index],
+            )
+            self._validate_list_uniqueness(attribute, value)
+
+    # --- Utility methods ---
+    @staticmethod
+    def _validate_array_shape(
+        attribute: attrs.Attribute, value: np.ndarray, expected_shape: tuple
+    ):
+        """Raise ValueError if the value does not have the expected shape."""
+        if value.shape != expected_shape:
+            raise logger.error(
+                ValueError(
+                    f"Expected '{attribute.name}' to have shape "
+                    f"{expected_shape}, but got {value.shape}."
+                )
+            )
+
+    @staticmethod
+    def _validate_list_length(
+        attribute: attrs.Attribute, value: list | None, expected_length: int
+    ):
+        """Raise a ValueError if the list does not have the expected length."""
+        if value is not None and len(value) != expected_length:
+            raise logger.error(
+                ValueError(
+                    f"Expected '{attribute.name}' to have "
+                    f"length {expected_length}, but got {len(value)}."
+                )
+            )
+
+    @staticmethod
+    def _validate_list_uniqueness(
+        attribute: attrs.Attribute, value: list | None
+    ):
+        """Raise a ValueError if the list does not have unique elements."""
+        if value is not None and len(value) != len(set(value)):
+            raise logger.error(
+                ValueError(
+                    f"Elements in '{attribute.name}' are not unique. "
+                    f"There are {len(value)} elements in the list, but "
+                    f"only {len(set(value))} are unique."
+                )
+            )
+
+    @classmethod
+    def validate(cls, ds: xr.Dataset) -> None:
+        """Validate the input as a valid ``movement`` dataset.
```

Comment:
Maybe this could be more explicit: "check the input dataset has required variables and dimensions"